### PR TITLE
  Fix unresponsive 'Add Record' action and snapping active when adding records to a non-spatial table

### LINKED ISSUE #55092: Unresponsive "Add record" button in non-geometry layer editing
### What is the bug or the crash?

When editing a non-geometry layer (table-only) in QGIS, the "Add record" button in the main interface becomes unresponsive after accepting or canceling changes in the corresponding form. The button remains in a pressed state, preventing users from clicking it again to add a new record. To add a new element to the table, users have to cancel the editing process and re-activate it. Only then can the "Add record" button be clicked again. 

### Steps to reproduce the issue

- Create a new project in QGIS.
- Add a non-geometry layer (e.g., Temporal, Geopackage, Shapefile).
- Activate the editing mode for the layer.
- Click the "Add record" button in the interface.
- If the layer has attributes, accept or cancel the changes made in the form.
- Notice that the "Add record" button remains pushed and unclickable.
imagen

### Versions

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.0//EN" "http://www.w3.org/TR/REC-html40/strict.dtd">
<html><head><meta http-equiv="Content-Type" content="text/html; charset=utf-8" /></head><body>

Versión de QGIS | 3.34.0-Prizren | Revisión del código de QGIS | ffbdd67881
-- | -- | -- | --
Versión Qt | 5.15.3
Versión de Python | 3.9.5
Versión de GDAL/OGR | 3.7.2
Versión de PROJ | 9.3.0
Versión del registro de base de datos EPSG | v10.094 (2023-08-08)
Versión de GEOS | 3.12.0-CAPI-1.18.0
Versión de SQLite | 3.41.1
Versión de PDAL | 2.5.5
Versión del cliente de PostgreSQL | 15.2
Versión de SpatiaLite | 5.1.0
Versión de QWT | 6.1.6
Versión de QScintilla2 | 2.13.4
Versión del SO | Windows 10 Version 2009
  |   |   |  
Complementos activos de Python
coordinate_capture | 0.2
DigitizingTools | 1.5.2
MemoryLayerSaver | 5.0.1
processing_fusion | 3.5
processing_saga_nextgen | 1.0.0
profiletool | 4.2.6
qad | 3.0.4
qgis_resource_sharing | 1.0.0
quick_map_services | 0.19.33
searchlayers | 3.0.14
SLD4raster | 1.0
Spanish_Inspire_Catastral_Downloader | 2.0
StreetView | 3.2
VectorBender | 0.2.1
wfsclient | 0.9.11
db_manager | 0.1.20
otbprovider | 2.12.99
processing | 2.12.99

</body></html>

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

This issue was not present prior to at least QGIS version 3.30.1, as far as I know. The first version in which it appeared was 3.32.

### LINKED ISSUE #56152: Add a new record in a table enable the snapping tool
### What is the bug or the crash?

After adding a new record in a table, the snapping tool is enabled

### Steps to reproduce the issue

1. Insert a table layer (without geometry)
2. Activate the edit mode for the layer
3. Turn on pan navigation (to highlight the existing issue)
4. Click on "Add a new record"
5. Confirm or close the form
6. Observe that the snapping tool becomes active (the cursor persists as the pan tool, even though panning functionality is not enabled)

*(The snapping square displayed in the animation should be purple instead of white)*

snapping

### Versions

3.28.13

### Supported QGIS version

- [X] I'm running a supported QGIS version according to [the roadmap](https://www.qgis.org/en/site/getinvolved/development/roadmap.html#release-schedule).

### New profile

- [X] I tried with a new [QGIS profile](https://docs.qgis.org/latest/en/docs/user_manual/introduction/qgis_configuration.html#working-with-user-profiles)

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by benwirf**
Fix 'Add Record' action remains toggled/not enabled and snapping active when adding records to a non-spatial table.
Fix open hand cursor persists when switching between non-spatial and spatial layers in edit mode.

Fixes https://github.com/qgis/QGIS/issues/55092
Fixes https://github.com/qgis/QGIS/issues/56152

## Description

This PR fixes two reported bugs and also fixes some undesirable/confusing UX/ UI behaviour in relation to the digitizing cursor when switching between non-spatial & spatial vector layers (all in edit mode) in the TOC.

- The cursor used when the Add Feature tool was active when adding records to a non-spatial layer was identical to the Pan cursor, but not responsive- potentially leading to confusion/ frustration. This PR replaces it with a generic Qt::ArrowCursor.

- When switching from a non-spatial table layer to a spatial vector layer with both in edit mode and the Add Feature tool active, the canvas/ map tool cursor remains set to the open hand cursor, making accurate digitizing more difficult. This PR also fixes that behavior.

This approach implements most of the logic in the `QgsMapToolDigitizeFeature`,  `QgsMapToolCapture` and `QgsMapToolAdvancedDigitizing` by overriding the canvas event methods and is based on discussions in a previous PR, #56283, for the same bug fixes.

### CODE REVIEW

**Comment by elpaso** [comment]
See errors here: https://github.com/qgis/QGIS/actions/runs/7969161837/job/21754535061?pr=56416#step:4:37

**Comment by agiudiceandrea** [comment]
> See errors here

Also SIP files not updated https://github.com/qgis/QGIS/actions/runs/7969161837/job/21754535704?pr=56416
See https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html#generating-the-sip-file (@elpaso is it the correct way to fix it?)

**Comment by benwirf** [comment]
Hi @agiudiceandrea and @elpaso. I have fixed the layout spacing and run the sipify script. Hopefully those tests will pass now. Thanks again for the hints. I need to properly set up the pre-commit hook for future PRs.
Cheers

**Comment by benwirf** [comment]
Hi @agiudiceandrea and @elpaso. I fixed the layout spacing and ran the sipify script on the `qgsmaptooldigitizefeature.h` header file. Unfortunately the sip check is still failing- I'm not sure why. I will try again tomorrow.

**Comment by agiudiceandrea** [comment]
@benwirf, to properly fix the sip check, just remove the python/PyQt6/gui/auto_generated/qgsmaptooldigitizefeature.sip.in and python/gui/auto_generated/qgsmaptooldigitizefeature.sip.in files in your branch fix_55092_56152_add_record_snapping_cursor and rename python/PyQt6/gui/auto_generated/qgsmaptooldigitizefeature.sip to python/PyQt6/gui/auto_generated/qgsmaptooldigitizefeature.sip.in and python/gui/auto_generated/qgsmaptooldigitizefeature.sip to python/gui/auto_generated/qgsmaptooldigitizefeature.sip.in

**Comment by elpaso** [comment]
> > See errors here
> 
> Also SIP files not updated https://github.com/qgis/QGIS/actions/runs/7969161837/job/21754535704?pr=56416 See https://docs.qgis.org/testing/en/docs/developers_guide/codingstandards.html#generating-the-sip-file (@elpaso is it the correct way to fix it?)

I usually run `scripts/sipify_all.sh`

**Comment by benwirf** [comment]
Thanks again for the hints guys, the sip checks are now passing. I am very grateful for your help. Cheers.

**Comment by agiudiceandrea** [comment]
I think the two ´*.sip´ files should be removed.

**Comment by benwirf** [comment]
@agiudiceandrea, oops- sorry for missing that. Done now, Cheers.

**Comment by agiudiceandrea** [comment]
hi @uclaros, @elpaso, it looks like all the tests are passing now for this PR, which fixes some bugs.

File: src/app/qgisapp.cpp [hunk 1]
```diff
@@ -10001,7 +10001,23 @@ void QgisApp::reshapeFeatures()
 
 void QgisApp::addFeature()
 {
-  mMapCanvas->setMapTool( mMapTools->mapTool( QgsAppMapTools::AddFeature ) );
+  if ( mMapCanvas->mapTool()->toolName() != tr( "Add feature" ) )
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:10004]:
```suggestion
  if ( mMapCanvas->mapTool() != mMapTools->mapTool( QgsAppMapTools::AddFeature ) )
```

That's safer then relying on a translated string.

File: src/app/qgisapp.cpp [hunk 2]
```diff
@@ -10001,7 +10001,23 @@ void QgisApp::reshapeFeatures()
...
+  {
+    mMapCanvas->setMapTool( mMapTools->mapTool( QgsAppMapTools::AddFeature ) );
+    if ( !mMapCanvas->currentLayer()->isSpatial() )
```
Comment (nyalldawson) [inline src/app/qgisapp.cpp:10007]: This doesn't look very clean to me. Wouldn't we be better just NOT setting the map tool to add feature?

**Comment by benwirf** [comment]
Hi @nyalldawson,
Thanks, good call on both suggestions. I have cleaned up that map tool check & `addFeature()` method logic. Hopefully it looks better now.
Cheers

**Comment by agiudiceandrea** [comment]
Another related issue report: https://github.com/qgis/QGIS/issues/57108.

**Comment by benwirf** [comment]
I think this is still just waiting for a review.

**Comment by agiudiceandrea** [comment]
Hi @elpaso, are you looking into this?
See another related issue report https://github.com/qgis/QGIS/issues/57464.

**Comment by agiudiceandrea** [comment]
Hi @nyalldawson, thi PR is still awaiting to be approved and merged.

**Comment by agiudiceandrea** [comment]
Another issue report: https://github.com/qgis/QGIS/issues/57995.

**Comment by agiudiceandrea** [comment]
@nyalldawson , any news on this PR?

**Comment by agiudiceandrea** [comment]
@nyalldawson uclaros, any news on this PR?

**Comment by agiudiceandrea** [comment]
Hi @benwirf, a fix for https://github.com/qgis/QGIS/issues/55092 has been recently proposed https://github.com/qgis/QGIS/pull/58680 and merged.
Do you think to create a new PR to also address https://github.com/qgis/QGIS/issues/56152?

**Comment by benwirf** [comment]
Hi @agiudiceandrea,

Thank you for letting me know. I honestly felt that this PR was thorough and comprehensive and fixed several reported bugs in a considered way- also addressing the related snapping & cursor issues etc.

I thought that the last commit back in March addressed the requested changes from @nyalldawson on this branch and @uclaros from #56283 and that it could be merged in due time. If there was something else which was blocking this PR from being approved and merged I would have gladly worked on fixing it over the intervening 6 months.

Maybe I was not proactive enough in actively pinging people myself for reviews but I'm very conscious of not seeming too demanding on others' time since I know everyone is very busy. Also, I was seeing that you were regularly reopening the PR and trying to keep it alive (thank you!).

Perhaps I should also have kept this branch up to date with master by rebasing more often. I am not sure, but I would genuinely like to know what I could do differently to avoid this in the future.

For me this is a bit of a disappointing situation and I'm feeling a bit emotional and raw about it right now, but I guess that to move forward, I will test the current master branch then try to pick out the remaining parts of this branch which I think are useful and open a new PR.

**Comment by benwirf** [comment]
I have closed this and opened #58755.

**Pull request closed without merging**